### PR TITLE
Added a compilation order heuristic for compiling C before Nc processes

### PR DESCRIPTION
**Pull request by srrisbud**
This heuristic takes effect only when both C and Nc Processes belong to the same ProcGroup

## Pull request checklist
Your PR fulfills the following requirements:
- [x] PR conforms to [Coding Conventions](https://lava-nc.org/developer_guide.html#coding-conventions)
- [x] [PR applys BSD 3-clause or LGPL2.1+ Licenses](https://lava-nc.org/developer_guide.html#add-a-license) to all code files
- [x] Lint (`flakeheaven lint src/lava tests/`) and (`bandit -r src/lava/.`) pass locally
- [x] Build tests (`pytest`) passes locally

## Pull request type
Please check your PR type:
- [x] Feature

### CODE REVIEW

**Review by PhilippPlank — APPROVED** [review]

**Review by mgkwill — APPROVED** [review]

**Review by mgkwill — APPROVED** [review]

File: src/lava/magma/compiler/compiler.py [hunk 1]
```diff
@@ -299,13 +299,42 @@ def _create_subcompilers(
...
+            # Remember the index for C and Nc subcompilers:
+            if isinstance(compiler, CProcCompiler):
+                c_idx.append(idx)
+            if isinstance(compiler, NcProcCompiler):
+                nc_idx.append(idx)
+
+        # Implement the heuristic "C-first Nc-second" whenever C and Nc
+        # compilers are in the list. The subcompiler.compile is called in
+        # their order of appearance in this list.
+        # ToDo: In the implementation below, we assume that there is only one
```
Comment (mathisrichter) [inline src/lava/magma/compiler/compiler.py:320]: That assumption should not be made. The compiler was designed to create multiple instances of the same type of subcompiler, one instance for every group of Processes.

File: src/lava/magma/compiler/compiler.py [hunk 2]
```diff
@@ -299,13 +299,42 @@ def _create_subcompilers(
...
         # Go through all required subcompiler classes...
-        for subcompiler_class, procs in compiler_type_to_procs.items():
+        for idx, (subcompiler_class, procs) in \
+                enumerate(compiler_type_to_procs.items()):
             # ...create the subcompiler instance...
             compiler = subcompiler_class(procs, self._compile_config)
             # ...and add it to the list.
             subcompilers.append(compiler)
-
+            # Remember the index for C and Nc subcompilers:
```
Comment (mathisrichter) [inline src/lava/magma/compiler/compiler.py:311]: If we want to keep the reordering here, please extract this entire reordering into a method with a meaningful method name like `_order_subcompilers_by_type` or something more fitting. This code, using indices, is at too low of an abstraction level to be placed inside of here directly.

**Review by mathisrichter — COMMENTED** [review]
Please consider whether this PR needs to be redacted and my comments addressed.

File: src/lava/magma/compiler/compiler.py (hunk 1 above)
Comment (mathisrichter) [reply src/lava/magma/compiler/compiler.py:320]: Sorry, this is not correct, I didn't see that this method was working on individual ProcGroups already.

**Comment by mathisrichter** [comment]
Please disregard my review.

**Pull request merged**